### PR TITLE
Support decorators + fix issue related to custom props in functions

### LINKED ISSUE #53: Decorator support + Missing defaults
Thanks for the extraction utilities. They are highly useful.

While benchmarking `extract-react-types`, I noticed a couple of issues:

* Decorators were disabled (Babel 7 change)?
* `addDefaultProps` could crash. If you want, I can figure out a test case for this.

Here are my initial fixes: https://github.com/bebraw/extract-react-types/commit/42fbdb625b447ea42a68e38fea7b6707441e8a56 . I can also PR if you want.

### PR DESCRIPTION
**Pull request by bebraw**
I noticed these corner cases in my own work. Likely you should add a test for Flow/decorators combination too.

Closes #53.

### CODE REVIEW

**Comment by bebraw** [comment]
I can see

```
Cannot call createBabylonOptions with object literal bound to opts because array
literal [1] is incompatible with string [2] in array element of property
plugins.
```

in the CI. Any idea what might be going on? Some version conflict?

**Comment by bebraw** [comment]
It's a Flow error, I see now.

**Comment by bebraw** [comment]
I added additional fix related to names of field lookups.

If someone knows Flow well, please check the CI failure. Maybe we just have to adjust Flow definitions for Babel somehow?

File: packages/extract-react-types/src/index.js [hunk 1]
```diff
@@ -1426,7 +1433,7 @@ function getContext(
   filename /*:? string */,
   resolveOptions /*:? Object */
 ) {
-  let plugins = ['jsx'];
+  let plugins = ['jsx', ['decorators', { decoratorsBeforeExport: true }]];
```
Comment (Noviny) [inline packages/extract-react-types/src/index.js]: Hey @bebraw, the flow error you're seeing is coming from this line. The reason is where it is defined in the library we are using [here](https://github.com/babel-utils/babylon-options/blob/daa7f9517623ccc43c2105e043cfdecbdff59444/index.js#L42), it says that plugins must be an array of strings, not recognising this syntax.

If you can add a `/* $FlowFixMe - need to update types in babylon-options */` to the line where plugin is passed in, you should be able to get flow passing, and then we can get this merged in.

Otherwise, these changes look fine.

**Comment by Noviny** [comment]
Left a comment on how to fix this. In addition, it would be great if you could run `yarn changeset` to add changeset information alongside the change you are making. :)

**Comment by bebraw** [comment]
I added one more fix (`key` was missing in a certain case) and did your flow fix thanks.

Looks like changesets get generated behind `.changeset` at the project root. Should I commit that? Here's a message for the changeset anyway (likely a patch release but you can do minor too):

```
- Decorators should work again
- Allow default props to be missing
- Return correct names for nested properties in an object
- Add `key` field to arrays (missing previously)
```

**Comment by Noviny** [comment]
Changesets are indeed a thing designed to be committed! There's some docs about them [here](https://github.com/atlassian/extract-react-types/blob/master/.changeset/README.md) and I'm trying to get better git support for them [here](https://github.com/Noviny/changesets/issues/5) as they currently don't integrate well on github into the review process.

For now, I'm going to merge this and will add your changeset, just so we can keep moving. Might try and get the other PR merged before doing a release.
